aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -501,14 +522,20 @@ def handle_known_unrecoverable_failure(self, calculation):
         PwCalculation.exit_codes.ERROR_OUT_OF_WALLTIME,
     ])
     def handle_out_of_walltime(self, calculation):
-        """Handle `ERROR_OUT_OF_WALLTIME` exit code: calculation shut down neatly and we can simply restart."""
+        """Handle `ERROR_OUT_OF_WALLTIME` exit code.
+
+        In this case the calculation shut down neatly and we can simply restart. We consider two cases:
+
+        1. If the structure is unchanged, we do a full restart.
+        2. If the structure has changed during the calculation, we restart from scratch.
```

Comment:
Since this corresponds to a clean exit, wouldn't we be able to reuse the charge density at least, even if the structure was changed? I know you didn't change anything in the logic, but I think there is no reason _not_ to reuse the charge density, is there?

Reply:
Myyyes, concur with your reasoning, I do. See the final two lines of [this comment](https://github.com/aiidateam/aiida-quantumespresso/pull/722#issuecomment-916835694). I also think that for `handle_relax_recoverable_ionic_convergence_error` we can restart from the charge density.

Reply:
Would the fact that the cell might have changed be a potential problem? Or since the charge density would also have been written on the new grid, that wouldn't matter? Not sure here

Reply:
But the charge density is just a 3 dimenional scalar function right? The issue with the cell changing is that the basis set is no longer complete, but I would say that this gets regenerated anyways and then the input charge density is expressed in this new basis set? 🤔

Reply:
Should be a scalar yes, but it is defined on a grid that relates to the crystal cell. If that changes but the grid of the charge density doesn't in a comparable way, they are effectively out of sync, no? However, I agree with your statement further down that we maybe not change this behavior now and keep what was there. Then we can address this in a follow-up PR and investigate it properly first